File: CTCI4_TreesGraphs/minimalTree.java
/*
Minimal Tree: Given a sorted (increasing order) array with unique integer elements, write an algorithm
to create a binary search tree with minimal height. 

SOLUTION: recursive! go through left and right and keep splitting up array
*/

Node minimalTreeRoot(int[] elements) {
    return minimalTree(elements, 0, elements.length-1);
}

Node minimalTree(int[] elements, int start, int end) {
    if(end < start) {
        return null;
    }
    int mid = (start+end)/2;

    Node n = new Node();
    n.left = minimalTree(elements, start, mid-1);
    n.right = minimalTree(elements, mid+1, end);
    n.data = elements[mid];
    return n;
}

File: CTCI4_TreesGraphs/minimalTree.py
'''
Minimal Tree: Given a sorted (increasing order) array with unique integer elements, write an algorithm
to create a binary search tree with minimal height. 

SOLUTION:
1. Recursive - make a minimal subtree from left and right of array, recursively
PROCESS:
1. Insert into the tree the middle element of the array.
2. Insert (into the left subtree) the left subarray elements.
3. Insert (into the right subtree) the right subarray elements. 
4. Recurse.
'''

def createMinimalSubtree(elements) {
    root = elements[len(elements)/2]
    root.left = createMinimalSubtree(elements[0, len(elements)/2])
    root.right = createMinimalSubtree(elements[(len(elements)/2)+1:])
}
